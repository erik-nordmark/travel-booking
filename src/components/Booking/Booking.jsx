import React, { useState, useEffect, useContext } from 'react';
import { useHistory } from 'react-router-dom';
import { CartContext } from 'contexts/CartContext';
import { DarkButton } from 'components/Button/Button';

import styles from './Booking.module.scss';

export const Booking = ({ title, price, unit }) => {
	const history = useHistory();

	const [adults, setAdults] = useState(0);
	const [children, setChildren] = useState(0);
	const [total, setTotal] = useState(0);

	const { updateCart } = useContext(CartContext);

	const addAdults = () => {
		setAdults((prevAdults) => prevAdults + 1);
	};

	const removeAdults = () => {
		setAdults((prevAdults) => {
			if (prevAdults > 0) {
				return prevAdults - 1;
			}

			return 0;
		});
	};

	const addChildren = () => {
		setChildren((prevAdults) => prevAdults + 1);
	};

	const removeChildren = () => {
		setChildren((prevChildren) => {
			if (prevChildren > 0) {
				return prevChildren - 1;
			}

			return 0;
		});
	};

	const bookExperience = () => {
		const booking = {
			name: title,
			adults,
			children,
			totalPrice: total,
		};

		updateCart((cart) => [...cart, booking]);
		history.push('/');
	};

	useEffect(() => {
		setTotal(adults * price + children * price);
	});

	return (
		<div className={styles.bookingWrapper}>
			<h3>{title}</h3>
			<p>{`from ${price} ${unit}`}</p>
			<p>Adult</p>
			<div className={styles.shopWrapper}>
				<button
					label='Remove adult'
					type='button'
					className={styles.remove}
					onClick={removeAdults}
				/>
				<input
					value={adults}
					onChange={(e) => setAdults(parseInt(e.currentTarget.value, 10))}
				/>
				<button
					label='Add adult'
					type='button'
					className={styles.add}
					onClick={addAdults}
				/>
			</div>

			<p>Children</p>
			<div className={styles.shopWrapper}>
				<button
					label='Remove children'
					type='button'
					className={styles.remove}
					onClick={removeChildren}
				/>
				<input
					value={children}
					onChange={(e) => setChildren(parseInt(e.currentTarget.value, 10))}
				/>
				<button
					label='Add children'
					type='button'
					className={styles.add}
					onClick={addChildren}
				/>
			</div>

			<p className={styles.total}>{`Total: Â£${total}`}</p>

			<DarkButton
				disabled={!adults && !children}
				label='Book experience'
				onClick={bookExperience}
			/>
		</div>
	);
};
