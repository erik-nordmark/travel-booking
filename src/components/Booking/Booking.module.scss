.bookingWrapper {
	width: 100%;
	min-width: 20vw;
	height: 100%;
	display: flex;
	flex-direction: column;
	text-align: left;

	h3 {
		text-align: left;
		font-family: Muli;
		font-style: normal;
		font-weight: bold;
		font-size: 18px;
		line-height: 23px;
		letter-spacing: 1.63636px;

		color: #000000;
	}

	.shopWrapper {
		display: inline-flex;

		button {
			border: none;
		}

		button:focus {
			outline: none;
		}

		input {
			width: 80px;
			margin: 0 5px;
		}

		.remove {
			width: 40px;
			height: 40px;
			background: #000000;
			display: flex;
			justify-content: center;
			align-items: center;
			padding: 0;
			margin: 0;
		}

		.remove:after {
			color: #ffffff;
			content: '-';
		}

		.add {
			width: 40px;
			height: 40px;
			background: #000000;
			display: flex;
			justify-content: center;
			align-items: center;
			padding: 0;
			margin: 0;
		}
		.add:after {
			color: #ffffff;
			content: '+';
		}
	}

	.total {
		margin: 3rem 0;
		font-family: Muli;
		font-style: normal;
		font-weight: bold;
		font-size: 14px;
		line-height: 18px;
		/* identical to box height, or 129% */

		letter-spacing: 1.27273px;

		color: #222222;
	}
}
