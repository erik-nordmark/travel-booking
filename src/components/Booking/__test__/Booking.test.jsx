import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { Booking } from 'components/Booking/Booking';

test('renders without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<Booking />, div);
});

test('matches snapshot', () => {
	const tree = renderer
		.create(<Booking titel='test' price={5} unit='per person' />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
