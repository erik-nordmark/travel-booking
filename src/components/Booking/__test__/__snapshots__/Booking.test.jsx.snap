// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="bookingWrapper"
>
  <h3 />
  <p>
    from 5 per person
  </p>
  <p>
    Adult
  </p>
  <div
    className="shopWrapper"
  >
    <button
      className="remove"
      label="Remove adult"
      onClick={[Function]}
      type="button"
    />
    <input
      onChange={[Function]}
      value={0}
    />
    <button
      className="add"
      label="Add adult"
      onClick={[Function]}
      type="button"
    />
  </div>
  <p>
    Children
  </p>
  <div
    className="shopWrapper"
  >
    <button
      className="remove"
      label="Remove children"
      onClick={[Function]}
      type="button"
    />
    <input
      onChange={[Function]}
      value={0}
    />
    <button
      className="add"
      label="Add children"
      onClick={[Function]}
      type="button"
    />
  </div>
  <p
    className="total"
  >
    Total: Â£0
  </p>
  <button
    className="button dark"
    data-testid="button"
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "width": "",
      }
    }
    type="button"
  >
    Book experience
  </button>
</div>
`;
